src/JlWrap/C.jl

```diff
@@ -339,22 +341,29 @@ end
 
 PyJuliaValue_IsNull(o) = Base.GC.@preserve o UnsafePtr{PyJuliaValueObject}(C.asptr(o)).value[] == 0
 
-PyJuliaValue_GetValue(o) = Base.GC.@preserve o PYJLVALUES[UnsafePtr{PyJuliaValueObject}(C.asptr(o)).value[]]
+PyJuliaValue_GetValue(o) = Base.GC.@preserve o begin
+    idx = UnsafePtr{PyJuliaValueObject}(C.asptr(o)).value[]
+    Base.@lock PYJLVALUES PYJLVALUES[].values[idx]
+end
 
 PyJuliaValue_SetValue(_o, @nospecialize(v)) = Base.GC.@preserve _o begin
     o = C.asptr(_o)
     idx = UnsafePtr{PyJuliaValueObject}(o).value[]
     if idx == 0
-        if isempty(PYJLFREEVALUES)
-            push!(PYJLVALUES, v)
-            idx = length(PYJLVALUES)
-        else
-            idx = pop!(PYJLFREEVALUES)
-            PYJLVALUES[idx] = v
+        Base.@lock PYJLVALUES begin
+            if isempty(PYJLVALUES[].free_slots)
+                idx = PYJLVALUES[].next_slot[]
+                PYJLVALUES[].next_slot[] += 1
+            else
+                idx = pop!(PYJLVALUES[].free_slots)
+            end
+            PYJLVALUES[].values[idx] = v
```

Comment:
This also seemed like a good thing to have... and might be relevant for why the segfault disappeared.